Review, please

supressionstop opened the pull request:
Branch created from initial commit.
Pulling master to this branch.

Commit:
config and logger

Commit:
logger improvements

Commit:
config improved

Commit:
http server

Commit:
storage, usecases, workers

Commit:
separated tables for each sport

Commit:
grpc

Commit:
dockerized processor, grpc server deferred start, changes

Commit:
tests, fixes

Commit:
linters, fixes

Commit:
docs, fixes

Commit:
rem to simplify review

Commit:
Revert "rem to simplify review"

This reverts commit eb6c7bfefc6f942b88de59c87190f383d31b273f.

xenking commented on Makefile:
```diff
@@ -1,13 +1,35 @@
+
+# helpers
+
+.PHONY: proto
+proto:
+	protoc --proto_path=api \
+			--go_out=. \
+			--go-grpc_out=. \
+			lines.proto
+
+.PHONY: mock
+mock:
+	mockery --config ./config/mockery.yml
```
We can create another command called `make generate`
```makefile
...
generate: proto mock
...

```

supressionstop replied:
TODO: make instruction "generate" to unite "proto" and "mock"

xenking commented on cmd/grpc_client/grpc_client.go:
```diff
@@ -0,0 +1,115 @@
+	"errors"
+	"flag"
+	"fmt"
+	"github.com/supressionstop/xenking_test_1/internal/server/pb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"io"
+	"log"
+	"os"
+	"strconv"
+	"strings"
```
Additional linter like `go-import` or better `gci` should be applied for proper imports order

supressionstop replied:
TODO: add linter gci to linter list

xenking commented on cmd/processor/processor.go:
```diff
@@ -0,0 +1,122 @@
+)
+
+func main() {
+	ctx := context.Background()
```
The app should handle OS signals. It should support graceful shutdown at the main func, not in the run(ctx). The function should be pure if possible and have only one purpose. In your code run(ctx) is the replacement for main() func

supressionstop replied:
TODO: remove run(), move behaviour in main().

My reason to use run() is in returning error, instead of exiting app.
The idea of run() function came out of this article:
https://grafana.com/blog/2024/02/09/how-i-write-http-services-in-go-after-13-years/#func-main-only-calls-run
And this course:
https://www.usegolang.com/

xenking commented on cmd/processor/processor.go:
```diff
@@ -0,0 +1,122 @@
+	ctx, stop := signal.NotifyContext(ctx, syscall.SIGINT, syscall.SIGTERM)
+	defer stop()
+
+	cfg, err := config.MustSetup(os.Getenv("APP_ENV"))
```
nit: function names with Prefix `Must` should not return an error and should call panic if the error happens inside the constructor.

supressionstop replied:
TODO: fix.

xenking commented on cmd/processor/processor.go:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	// provider
+	httpClient := &http.Client{
+		Transport:     nil,
+		CheckRedirect: nil,
+		Jar:           nil,
+		Timeout:       cfg.Provider.HTTPTimeout,
+	}
+
+	kiddy, err := provider.NewKiddy(httpClient, cfg.Provider.BaseURL)
+	if err != nil {
+		return fmt.Errorf("failed to setup provider: %w", err)
+	}
```
We can move this provider init part to a separate constructor

supressionstop replied:
TODO: inline.

xenking commented on cmd/processor/processor.go:
```diff
@@ -0,0 +1,122 @@
+
+	logger.Info("running")
+	select {
+	case err := <-httpServer.ErrChan:
+		if err != nil {
+			logger.Error("http server error", slog.Any("err", err))
+		}
+
+		ctx.Done()
+	case err := <-grpcServer.ErrChan:
+		if err != nil {
+			logger.Error("grpc server error", slog.Any("err", err))
+		}
+
+		ctx.Done()
+	case <-ctx.Done():
+		grpcServer.GracefulStop()
+
+		err := httpServer.Shutdown(ctx)
+		if err != nil {
+			logger.Error("failed to shutdown http server", slog.Any("err", err))
+		}
+
+		stop()
+	}
```
It's better to run a separate goroutine for each async worker like httpServer and grpcServer and handle ctx.Done() inside each worker. If you want to handle errors from workers and stop the app you should use panic instead.

supressionstop replied:
TODO: fix this.

xenking commented on docker/processor/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+WORKDIR /app
+COPY config config
+COPY --from=builder /app/processor ./processor
+CMD ./processor
```
nit: can use entrypoint instead

supressionstop replied:
TODO: use entrypoint.

xenking commented on internal/config/config.go:
```diff
@@ -0,0 +1,155 @@
+			return fmt.Errorf("cant parse worker idx: %s", name)
+		}
+
+		if len(nameParts) == 3 && nameParts[2] == "SPORT" {
```
Some complex parsing logic in config init. I'm unsure if this is the right place to install worker configurations. `SPORT`, `INTERVAL`, `POLL` can be replaced with constants for clarity

supressionstop replied:
TODO: refactor to external package after ddd.

xenking commented on internal/provider/kiddy.go:
```diff
@@ -0,0 +1,101 @@
+	path := "/v1/lines/"
+
+	fullUrl := p.baseUrl.JoinPath(path, sportName)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fullUrl.String(), nil)
```
nit: `nil` can be replaced with `http.NoBody` - as linter recommends

supressionstop replied:
TODO: fix.

xenking commented on internal/usecase/dto/line_map.go:
```diff
@@ -0,0 +1,16 @@
+	"github.com/supressionstop/xenking_test_1/internal/usecase/enum"
+)
+
+type LineMap map[enum.Sport]entity.Line
```
nit: `dto` pattern is not commonly used in Go, usually these entities belong to the repository level and we create struct converters from the `domain` (`entity`, `models`) level to repository entities. So in this case it's looks like we can rename to `domain` or `models` package

supressionstop replied:
TODO:
1. get rid of dto package;
2. move structs to models package.

xenking commented on internal/provider/kiddy.go:
```diff
@@ -0,0 +1,101 @@
+		return Line{}, fmt.Errorf("http code: %w", err)
+	}
+
+	bodyBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return Line{}, fmt.Errorf("read response body: %w", err)
+	}
+
+	var lines struct {
+		Lines Line `json:"lines"`
+	}
+	err = json.Unmarshal(bodyBytes, &lines)
+	if err != nil {
+		return Line{}, fmt.Errorf("unmarshall body: %w", err)
+	}
```
nit: better to replace with the streaming body. Please see json.NewDecoder().Decode(lines)

supressionstop replied:
TODO: fix.

xenking commented on internal/provider/kiddy.go:
```diff
@@ -0,0 +1,101 @@
+		return Line{}, fmt.Errorf("read response body: %w", err)
+	}
+
+	var lines struct {
+		Lines Line `json:"lines"`
+	}
```
nit: I know that it's ok to have an anonymous struct here but maybe we can move it outside this func and name it for clarity.

supressionstop replied:
TODO: fix.

xenking commented on internal/usecase/enum/sport.go:
We don't have enums in Go so we don't need a separate package for that. It should be somewhere in `entities` package

supressionstop replied:
TODO: extract enums to `entities`.

xenking commented on internal/usecase/dto/lines_diff.go:
```diff
@@ -0,0 +1,5 @@
+
+import "github.com/supressionstop/xenking_test_1/internal/usecase/enum"
+
+type LinesDiff map[enum.Sport]string
```
It's not worth to create a separate file only for one type. We can reduce the number of files and aggregate all declarations in one file if it's not too big.

supressionstop replied:
TODO:
1. extract dto;
2. minimize files count.

xenking commented on internal/server/grpc.go:
```diff
@@ -0,0 +1,115 @@
+	"google.golang.org/grpc/status"
+)
+
+type Client string
```
nit: Do we really need a separate type for the string here?

supressionstop replied:
No, we don't.
`Client` clarifies what grpc peer address is.

I would prefer to leave the Client as is.

TODO: unexport `Client`.

xenking commented on internal/worker/pool.go:
```diff
@@ -0,0 +1,78 @@
+	ErrChan        chan error
+}
+
+func NewPool(cfg config.Config, logger *slog.Logger, fetchLine usecase.FetchLine) *Pool {
```
From my understanding, this package can be a pure implementation for the worker pool without any domain logic leaks. I understand that in this implementation it's OK.

supressionstop replied:
TODO: try to replace many structs to worker pool pattern.

supressionstop replied:
Need to discuss new solution.

xenking commented on internal/worker/pool.go:
```diff
@@ -0,0 +1,78 @@
+	pool := &Pool{
+		logger:         logger,
+		workers:        make([]*Worker, len(cfg.Workers)),
+		FirstSyncChan:  make(chan struct{}),
+		workerSyncChan: make(chan struct{}, len(cfg.Workers)),
+		workersErrChan: make(chan error, len(cfg.Workers)),
```
It seems to me like an overcomplicated implementation of the worker's pool. I believe we can get rid of `FirstSyncChan`, `ErrChan`, `workersErrChan`, and `workersSyncChan`. We can log workers errors inside worker and remove a worker sync.

supressionstop replied:
TODO: try to replace many structs to worker pool pattern.

supressionstop replied:
Need to discuss new solution

xenking commented on internal/worker/pool.go:
```diff
@@ -0,0 +1,78 @@
+	logger         *slog.Logger
+	FirstSyncChan  chan struct{}
+	workerSyncChan chan struct{}
+	workersErrChan chan error
```
nit: It's a bad practice to use type names in variable names. We should always name variables without it's underlying type included.

supressionstop replied:
TODO: 
1. remove package or type prefixes;
2. try to find linter to highlight such problems.

supressionstop replied:
Removed and linters were added.

xenking commented on internal/usecase/interface.go:
It's a bad practice to create a separate file called `interfaces.go` or name something like `IConstructor` or `ConstructorInterface` or `constructorImpl`. We should write an interface definition at the place where we actually expect it and will be using it. So it's OK to have the same interface declaration in different files because we want to have a loose coupling.

supressionstop replied:
TODO: get rid of interfaces.go.

xenking commented on internal/server/grpc.go:
```diff
@@ -0,0 +1,115 @@
+	addr                string
+	grpcServer          *grpc.Server
+	logger              *slog.Logger
+	subscriptionManager *SubscriptionManager
+	pb.UnimplementedLinesServer
```
In the current implementation, this `server` package serves as a transport and controller handler. I believe this should be separated into a transport level (HTTP, grpc, or Async) and an actual controller level where we will use `SubscriptionManager`, `worker.Pool` etc. At a transport level, we should apply the ports approach to provide all user inputs to the controller level.

supressionstop replied:
TODO:
1. extract gRPC and HTTP into separate package - transport;
2. extract transport controllers into separate package - controllers

xenking commented on internal/usecase/calculate_diff.go:
```diff
@@ -0,0 +1,45 @@
+	"github.com/supressionstop/xenking_test_1/internal/usecase/dto"
+)
+
+type CalculateDiffUseCase struct{}
```
If we are already in `usecase` package we don't need to append the suffix `UseCase` at the type

supressionstop replied:
TODO: remove package suffixes.

xenking commented on tools/grpc_client:
We should not include compiled code in our repos if possible even if it serves testing purposes

supressionstop replied:
TODO: remove binary and tools folder.

Commit:
rem extra config logic
fix pb generation
remove usecase suffix, rem tools
remove usecase suffix
import fix
controllers; app,infra,usecase
separate controllers
rem interface.go
types removed from vars, linters added
rem chan suffixes
worker max error limit
workers refactored
client addr unexported
json decoder
rem extra packages
http.NoBody
docker entrypoint
http, grpc servers context and start handling
httpserver inlined in kiddy constructor
rem Must from config
rem run(), logic moved to main
gci linter
makefile generate

supressionstop commented:
Made changes due to review.
